staging/vhost-device-can/src/vhu_can.rs

```diff
@@ -0,0 +1,2175 @@
+        }
+
+        // Treat Vcan interface as CANFD if MTU is set to 64 bytes.
+        //
+        // Vcan can not be configured as CANFD interface, but it is
+        // possible to configure its MTU to 64 bytes. So if a messages
+        // bigger than 8 bytes is being received we consider it as
+        // CANFD message.
+        let can_name = self.controller.read().unwrap().can_name.clone();
+        if self.check_features(VIRTIO_CAN_F_CAN_FD)
+            && response.length.to_native() > 8
+            && can_name == "vcan0"
+        {
+            response.flags = (response.flags.to_native() | CAN_FRMF_TYPE_FD).into();
+            warn!("\n\n\nCANFD VCAN0\n\n");
+        }
+
+        // Adapt CAN length based on negotiated features
+        if (response.flags.to_native() & CAN_FRMF_TYPE_FD) != 0 {
+            can_rx.flags = (can_rx.flags.to_native() | VIRTIO_CAN_FLAGS_FD).into();
+            if response.length.to_native() > 64 {
+                warn!(
+                    "%s(): Cut length from {} to 64\n",
+                    response.length.to_native()
+                );
+                can_rx.length = 64.into();
+            }
+        } else if response.length.to_native() > 8 {
+            warn!(
+                "%s(): Cut length from {} to 8\n",
+                response.length.to_native()
+            );
+            can_rx.length = 8.into();
+        }
+
+        can_rx.sdu.copy_from_slice(&response.sdu[0..64]);
+        CanController::print_can_frame(can_rx);
+
+        Ok(can_rx)
+    }
+
+    fn process_ctrl_requests(
+        &self,
+        requests: Vec<CanDescriptorChain>,
+        vring: &VringRwLock,
+    ) -> Result<bool> {
+        if requests.is_empty() {
+            return Ok(true);
+        }
+
+        for desc_chain in requests {
+            let atomic_mem = self.mem.as_ref().unwrap().memory();
+
+            let mut reader = desc_chain
+                .clone()
+                .reader(&atomic_mem)
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            let mut writer = desc_chain
+                .clone()
+                .writer(&atomic_mem)
+                .map_err(|_| Error::DescriptorWriteFailed)?;
+
+            let request = reader
+                .read_obj::<VirtioCanCtrlRequest>()
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            // This implementation requires the CAN devices to be already in UP state
+            // before starting. This code does not trigger state changes [UP/DOWN] of
+            // the host CAN devices.
+            let response = match request.msg_type.into() {
+                VIRTIO_CAN_SET_CTRL_MODE_START => {
+                    if self.controller.write().unwrap().ctrl_state == CAN_CS_STARTED {
+                        VIRTIO_CAN_RESULT_NOT_OK
+                    } else {
+                        // TODO: Trigger a config_change notification
```

Comment:
If I understand correctly the spec, the device may omit the change notification (section 5.20.5.1):
```
As the configuration status change is caused by a request from the driver the device is
allowed to omit the configuration change notification here. The device marks the request used when the
CAN controller has finalized the transition to the requested controller mode.
```

Reply:
I believe that's correct, so in the control queue case the code can stay as it is. In the rest of the cases, where the device changes the config status to `BUSOFF` unsolicitedly,  I think we should trigger that notification.